### ISSUE TITLE
[12.0] EXR encode/export produces swizzled image content (ABGR instead of RGBA)

### ISSUE THREAD

**Issue by Froyok** [love2d/love#2070]
When exporting an EXR image from Love, the content of the file shows the wrong colors and transparency.
All the image reader I tried (XnView, Substance Designer, Krita, Tacent view) show the EXR file incorrectly as well, so to me this is Love not exporting/writing the EXR properly. After investigating a bit, it seems the RGBA channels are swizzled as ABGR. 

---

With the following example, I compensate the swizzle via the shader before hand which allows me to compensate the issue:

```lua
local ShaderVertex = [[
	#pragma language glsl4
	varying vec4 VarScreenPosition;
	varying vec2 VarVertexCoord;

	void vertexmain()
	{
		VarVertexCoord = vec2(
			(love_VertexID << 1) & 2,
			love_VertexID & 2
		);

		VarScreenPosition = vec4(
			VarVertexCoord.xy * vec2(2.0, -2.0) + vec2(-1.0, 1.0),
			0,
			1
		);

		gl_Position = VarScreenPosition;
	}
]]

local ShaderFragment = [[
	#pragma language glsl4
	varying vec2 VarVertexCoord;

	uniform float Swizzle;

	out vec4 FragColor;
	void pixelmain()
	{
		vec2 UV = VarVertexCoord.xy;

		vec4 Output = vec4(
			UV.x,
			UV.y,
			1.0 - UV.y,
			1.0
		);

		if( Swizzle > 0.0 )
		{
			Output = Output.abgr;
		}

		FragColor = Output;
	}
]]

function love.load( Args )
	Shader = love.graphics.newShader(
		ShaderFragment,
		ShaderVertex
	)

	Canvas = love.graphics.newCanvas(
		512, 512,
		{
			type 		= "2d",
			format 		= "rgba16f",
			mipmaps 	= "none"
		}
	)

	----------------------------------
	-- Render default version
	----------------------------------
	love.graphics.setShader( Shader )
	love.graphics.setBlendMode( "none" )

	love.graphics.setCanvas( Canvas )
	Shader:send( "Swizzle", 0.0 )
	love.graphics.drawFromShader( "triangles", 3, 1 )
	love.graphics.setCanvas()

	local Data = love.graphics.readbackTexture( Canvas )

	local File = love.filesystem.newFile( "output_default.exr" )
	File:open( 'w' )
	File:write( Data:encode("exr") )
	File:close()

	----------------------------------
	-- Render swizzled version
	----------------------------------
	love.graphics.setCanvas( Canvas )
	Shader:send( "Swizzle", 1.0 )
	love.graphics.drawFromShader( "triangles", 3, 1 )
	love.graphics.setCanvas()

	Data = love.graphics.readbackTexture( Canvas )

	File = love.filesystem.newFile( "output_swizzled.exr" )
	File:open( 'w' )
	File:write( Data:encode("exr") )
	File:close()

	love.event.quit( 0 )
end
```

---

Results (converted in PNG to be visible here):

- Default result: ![output_default](https://github.com/love2d/love/assets/6290387/97cd7d42-98bb-4efc-b761-1deeace26c6f)
- Swizzled result: ![output_swizzled](https://github.com/love2d/love/assets/6290387/99146737-332d-4cc0-bffd-146a0ef665e8)

---

System details:

- Love 12 (using this binary [8606538038](https://github.com/love2d/love/actions/runs/8606538038))
- OS: Linux Manjaro
- Arch: x64
- Renderer info: 
    - 4.6 (Core Profile) Mesa 24.0.2-manjaro1.1
    - AMD Radeon RX 7900 XT (radeonsi, navi31, LLVM 16.0.6, DRM 3.54, 6.6.26-1-MANJARO)

**Comment by Froyok** [comment]
I wonder if the issue is here: https://github.com/love2d/love/blob/657b3de9d9cd585d2ff73a196e3fc66a15eb6267/src/modules/image/magpie/EXRHandler.cpp#L310
Looking around on other implementations (relying on tinyexr like Love), they note:
`// Must be (A)BGR order, since most of EXR viewers expect this channel order.` ([from here](https://gitlab.inria.fr/alta/alta/-/blame/523ad1f33db210691d04efcea30fd1f5a944d932/sources/plugins/data_io/EXR_IO.h#L163), another [example here](https://github.com/cinder/Cinder/blob/e7ea5d98f64048fe0cd62d51cfa8459128021ee1/src/cinder/ImageFileTinyExr.cpp#L202)).

**Comment by slime73** [comment]
Apple's EXR decoder they use is working as expected on my Mac (the default and swizzled results are the opposite of your pic), so that comment is probably right. I bet most EXR viewers on Linux are using the same decoding library that isn't handling RGBA channel order correctly.

**Comment by Froyok** [comment]
Looks like Godot had the same issue (and in this case Blender and Gimp expected the swizzled result too): https://github.com/godotengine/godot/issues/55472

**Comment by Froyok** [comment]
I ran my example script on Windows and opened the exr resulting files in Substance Designer (Windows version too) and they show the same issue as in my original post.

I was pretty sure already that Designer wasn't relying on a system library like you mentioned, but I wanted to check just to be sure. On Designer we use FreeImage (which itself relies on OpenEXR).

I don't know what Apple is doing, but it seems to be going the opposite way of everybody (as usual :grin: ).

**Comment by slime73** [comment]
Yeah Apple often uses its own proprietary decoders rather than third party ones, for images. love is actually doing the right thing here as far as I know (it should even be decoding files correctly that display wrong in things that use FreeImage), but I'll still put in a workaround to make things display right when loaded with the broken decoders.

**Comment by Froyok** [comment]
According to OpenEXR doc:
`Channels are stored in alphabetical order, according to channel names. Within a channel, pixels are stored left to right.` (from [here](https://openexr.com/en/latest/OpenEXRFileLayout.html)).

So my guess on what is happening here is that the order in which the channels are written in the file is not ABGR, and most decoders just get the list of channels by order assuming that order (without looking at the actual channel name). If the Apple decoder doesn't do this assumption and properly check the name, then it explains why the result are inverted.

So I think the change should be quite simple here and require to just change the order (like they did on Godot).

**Comment by Froyok** [comment]
Another remark on this: would you consider exposing some parameters ?
Ideally it would be nice to specify the compression scheme. Saving a roughly 1024x512 pixels image on my side takes quite a few ms to do, I wonder if no compression could improve that (since in that case I don't care about the footprint).

**Comment by Froyok** [comment]
So I did more investigation and tests:
- I checked if Gimp was able to open EXR files from love. Gimp behave like many other applications out there and import the file incorrectly (swapped channels).
- [Cuttlefish](https://github.com/akb825/Cuttlefish), the library I use to convert EXRs into DDS files also don't open Love's EXR files properly (relies on FreeImage).
- I looked into how we import EXR files in Substance Painter, and in this case we allocate memory with explicit named R, G, B, and A memory buffers that we then pass to the OpenEXR library (we don't use FreeImage anymore for this format) and it's also not imported correctly.

I have no idea what Apple does, but if even with the help of the official OpenEXR library we cannot easily open EXR files then in my opinion it is simply broken as-is. :(
